File: app/src/main/java/com/example/myapplication/base/utils/callback/ItemTouchCallback.java
package com.example.myapplication.base.utils.callback;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.chad.library.adapter.base.BaseQuickAdapter;

public class ItemTouchCallback extends ItemTouchHelper.Callback {

    private final BaseQuickAdapter adapter;

    public ItemTouchCallback(BaseQuickAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
        int swipeFlags = 0;
        return makeMovementFlags(dragFlags, swipeFlags);
    }

    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
        int fromPosition = viewHolder.getBindingAdapterPosition();
        int toPosition = target.getBindingAdapterPosition();
        adapter.notifyItemMoved(fromPosition, toPosition);
        return true;
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
        // 可以处理滑动删除的逻辑
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return true; // 支持长按拖动
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return false; // 不支持滑动删除
    }
}

File: app/src/main/java/com/example/myapplication/module/recyclerview/drag/ItemTouchCallback.kt
package com.example.myapplication.module.recyclerview.drag

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

abstract class ItemTouchCallback : ItemTouchHelper.Callback() {

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ): Int {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
        val swipeFlags = 0
        return makeMovementFlags(dragFlags, swipeFlags)
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean {
        // 调用外部的 onMoveListener 逻辑
        onMoveListener(recyclerView, viewHolder, target)
        return true
    }

    /**
     * 抽象方法，由外部重写实现拖拽逻辑
     */
    abstract fun onMoveListener(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    )

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        // 可在子类中实现滑动删除逻辑
    }

    override fun isLongPressDragEnabled(): Boolean {
        return true // 支持长按拖动
    }

    override fun isItemViewSwipeEnabled(): Boolean {
        return false // 不支持滑动删除
    }
}
